cmd/terminal.go

```diff
@@ -0,0 +1,430 @@
+		if err != nil {
+			return errors.Wrap(err, "Unable to get the terminal size")
+		}
+		isTerminal = true
+	}
+
+	// connect to the websocket
+	conn, err := c.connect()
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	// set the terminal in raw mode
+	if isTerminal {
+		fmt.Fprintln(os.Stderr, "Press CTRL+] to quit the session")
+
+		oldState, err := term.MakeRaw(termID)
+		if err != nil {
+			return errors.Wrap(err, "Unable to set the terminal in raw mode")
+		}
+		defer func() {
+			_ = term.Restore(termID, oldState)
+		}()
+	}
+
+	// start the ping-pong connection health-check
+	go c.pingPong(ctx, conn)
+
+	// start the shell
+	if err := c.startShell(conn, termWidth, termHeight); err != nil {
+		return err
+	}
+
+	// message channel
+	msgChan := make(chan *ws.ProtoMsg)
+
+	c.running = true
+	go c.pipeStdin(msgChan, os.Stdin)
+	go c.pipeStdout(conn, os.Stdout)
+
+	// handle CTRL+C and signals
+	quit := make(chan os.Signal, 1)
+	signal.Notify(quit, unix.SIGINT, unix.SIGTERM)
+
+	// resize the terminal window
+	go c.resizeTerminal(ctx, msgChan, termID, termWidth, termHeight)
+
+	// wait for CTRL+C, signals or stop
+	for c.running {
+		select {
+		case msg := <-msgChan:
+			data, err := msgpack.Marshal(msg)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+				break
+			}
+			_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+			_ = conn.WriteMessage(websocket.BinaryMessage, data)
+		case <-quit:
+			c.running = false
+		case <-c.stop:
+			c.running = false
+		}
+	}
+
+	// cancel the context
+	cancelContext()
+
+	// stop shell message
+	if err := c.stopShell(conn); err != nil {
+		return err
+	}
+
+	// return the error message (if any)
+	return c.err
+}
+
+func (c *TerminalCmd) resizeTerminal(ctx context.Context, msgChan chan *ws.ProtoMsg, termID int, termWidth int, termHeight int) {
+	resize := make(chan os.Signal, 1)
+	signal.Notify(resize, syscall.SIGWINCH)
+	defer signal.Stop(resize)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-resize:
+			newTermWidth, newTermHeight, _ := terminal.GetSize(termID)
+			if newTermWidth != termWidth || newTermHeight != termHeight {
+				termWidth = newTermWidth
+				termHeight = newTermHeight
+				m := &ws.ProtoMsg{
+					Header: ws.ProtoHdr{
+						Proto:   ws.ProtoTypeShell,
+						MsgType: wsshell.MessageTypeResizeShell,
+						Properties: map[string]interface{}{
+							"terminal_width":  termWidth,
+							"terminal_height": termHeight,
+						},
+					},
+				}
+				msgChan <- m
+			}
+		}
+	}
+}
+
+func (c *TerminalCmd) Stop() {
+	c.running = false
+	c.stop <- struct{}{}
+}
+
+func (c *TerminalCmd) pipeStdin(msgChan chan *ws.ProtoMsg, r io.Reader) {
+	s := bufio.NewReader(r)
+	for c.running {
+		raw := make([]byte, 1024)
+		n, err := s.Read(raw)
+		if err != nil {
+			if c.running {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			} else {
+				c.Stop()
+			}
+			break
+		}
+		// CTRL+] or EOF, terminate the shell
```

Comment:
ASCII(4) is EOT - End of transmision
```suggestion
		// CTRL+] or EOT, terminate the shell
```

Reply:
ack